### PR TITLE
[REVIEW] Allow passing a dict in feat_name for add_edge_data and add_node_data

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR adds support for dict in feat_name for add_edge_data and add_node_data. 

- [x] Code Cleanup and ensure it lines with DGL 
- [x] Add tests
- [x] Fix a `get_feature_storage` bug when device is set

### CODE REVIEW

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 1]
```diff
@@ -54,9 +54,9 @@ def add_node_data(
         self,
         df,
         node_col_name,
-        feat_name=None,
         ntype=None,
-        is_single_vector_feature=True,
+        feat_name=None,
+        contains_vector_features=False,
     ):
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/graph_store.py:60]: Changed to this to keep the API cleaner and intuitive as you would want to provide feat_name together with contains_vector_features

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 2]
```diff
@@ -586,20 +570,17 @@ def fetch(self, indices, device=None, pin_memory=False, **kwargs):
...
+        cap = subset_df.to_dlpack()
+        tensor = self.from_dlpack(cap)
+        del cap
+        if device:
+            if not isinstance(tensor, cp.ndarray):
+                # Cant transfer to different device for cupy
                 tensor = tensor.to(device)
-            else:
-                return tensor
+        return tensor
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/graph_store.py:583]: Fixed an error here when a user provides device, cant test this as that needs Pytorch installed here.

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 3]
```diff
@@ -128,47 +115,34 @@ def add_edge_data(
...
-                self.edata_feat_col_d[col] = [col]
+        _update_feature_map(
+            self.edata_feat_col_d, feat_name, contains_vector_features, columns
+        )
 
         # Clear properties if set as data has changed
         self.__clear_cached_properties()
 
-    def get_node_storage(self, feat_name, ntype=None):
+    def get_node_storage(self, key, ntype=None, indices_offset=0):
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/graph_store.py:145]: Changed to key to match DGL.

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 4]
```diff
@@ -179,22 +153,23 @@ def get_node_storage(self, feat_name, ntype=None):
...
         return CuFeatureStorage(
             pg=self.gdata,
             columns=columns,
             storage_type="node",
+            indices_offset=indices_offset,
             backend_lib=self.backend_lib,
         )
 
-    def get_edge_storage(self, feat_name, etype=None):
+    def get_edge_storage(self, key, etype=None, indices_offset=0):
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/graph_store.py:172]: Changed to key to match DGL.

File: python/cugraph/cugraph/gnn/graph_store.py (hunk 3 above)
Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/graph_store.py:145]: See : https://github.com/dmlc/dgl/blob/2630d2eb0231cc9de2cb953e49d092ad556e3be9/python/dgl/heterograph.py#L1604-L1605

File: python/cugraph/cugraph/gnn/graph_store.py (hunk 4 above)
Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/graph_store.py:172]: See: https://github.com/dmlc/dgl/blob/2630d2eb0231cc9de2cb953e49d092ad556e3be9/python/dgl/heterograph.py#L1608-L1610

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 5]
```diff
@@ -548,6 +528,7 @@ def __init__(self, pg, columns, storage_type, backend_lib="torch"):
         self.storage_type = storage_type
 
         self.from_dlpack = from_dlpack
+        self.indices_offset = indices_offset
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/graph_store.py:531]: Setting offset makes upstream code cleaner

**Comment by VibhuJawa** [comment]
@gpucibot  rerun tests

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 6]
```diff
@@ -586,20 +570,17 @@ def fetch(self, indices, device=None, pin_memory=False, **kwargs):
...
-            tensor = self.from_dlpack(subset_df.to_dlpack())
 
-        if isinstance(tensor, cp.ndarray):
-            # can not transfer to
-            # a different device for cupy
-            return tensor
-        else:
-            if device:
+        if len(subset_df) == 0:
+            raise ValueError(f"indices = {indices} not found in FeatureStorage")
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/graph_store.py:575]: Ahh clever, did not know about this feature.

File: python/cugraph/cugraph/gnn/graph_store.py [hunk 7]
```diff
@@ -68,54 +68,41 @@ def add_node_data(
...
-            Whether to treat all the columns of the dataframe being added as
-            a single 2d feature
+        feat_name : {} or string
+            A map of feature names under which we should save the added
+            properties like {"feat_1":[f1, f2], "feat_2":[f3, f4]}
+            (ignored if contains_vector_features=False and the col names of
+            the dataframe are treated as corresponding feature names)
+        contains_vector_features : False
+            Wether to treat the columns of the dataframe being added as
+            as 2d features
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/graph_store.py:84]: We can have cases when we have nd features but we can not handle that with dataframes currently .  

We can probably figure out a generic mechanism then.   Maybe this is another motivation to move away from using dataframes to using arrays to store features

**Review by VibhuJawa — COMMENTED** [review]
Thanks @rlratzel , Have addressed your reviews and added comment on  the questions. Let me know what else needs to be done.

**Review by rlratzel — APPROVED** [review]

**Comment by VibhuJawa** [comment]
@gpucibot  merge

**Comment by VibhuJawa** [comment]
@gpucibot merge

**Comment by rlratzel** [comment]
@gpucibot merge

**Pull request merged**